### PR TITLE
feature: Add support for SageMaker lineage queries in action

### PR DESCRIPTION
**Pull request by yzhu0**
*Issue #, if available:*

*Description of changes:*
feature: Add support for SageMaker lineage queries in action.
- Get downstream Artifacts in aciton
- Get upstream Artifacts in action
- Get dataset in ModelPackageApprovalAction
- Get ModelPackage in ModelPackageApprovalAction
- Get Endpoint in ModelPackageApprovalAction

*Testing done:*
- unit test
- integration test

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/lineage/action.py [hunk 1]
```diff
@@ -250,3 +259,107 @@ def list(
             max_results=max_results,
             next_token=next_token,
         )
+
+    def upstream_artifacts(
+        self, direction: LineageQueryDirectionEnum = LineageQueryDirectionEnum.ASCENDANTS
```
Comment (staubhp) [inline src/sagemaker/lineage/action.py:264]: If we're going the route of having a separate upstream/downstream method for each query, we shouldn't allow the direction to be changed. 

I'm wondering if it's worth just having an `artifacts` method with direction defaulted to BOTH. Users can change to ASCENDANTS/DESCENDANTS based on their needs

Comment (staubhp) [reply src/sagemaker/lineage/action.py:264]: Either way, let's try to cut down on the code duplication if possible. Have a `_artifacts()` method that both of these call with the common code

File: src/sagemaker/lineage/action.py [hunk 2]
```diff
@@ -250,3 +259,107 @@ def list(
...
+        return [vertex.to_lineage_object() for vertex in query_result.vertices]
+
+
+class ModelPackageApprovalAction(Action):
+    """An Amazon SageMaker model package approval action, which is part of a SageMaker lineage."""
+
+    def upstream_datasets(
+        self, direction: LineageQueryDirectionEnum = LineageQueryDirectionEnum.ASCENDANTS
+    ) -> List[Artifact]:
+        """Use a lineage query to retrieve all upstream artifacts that use this action.
```
Comment (staubhp) [inline src/sagemaker/lineage/action.py:310]: "all upstream artifacts" -> "all upstream datasets"

File: src/sagemaker/lineage/action.py [hunk 3]
```diff
@@ -250,3 +259,107 @@ def list(
...
+        )
+        query_result = LineageQuery(self.sagemaker_session).query(
+            start_arns=[self.action_arn],
+            query_filter=query_filter,
+            direction=direction,
+            include_edges=False,
+        )
+        return [vertex.to_lineage_object() for vertex in query_result.vertices]
+
+    def model_package(self) -> List[Artifact]:
```
Comment (staubhp) [inline src/sagemaker/lineage/action.py:329]: This return type is wrong

File: src/sagemaker/lineage/action.py [hunk 4]
```diff
@@ -250,3 +259,107 @@ def list(
...
+
+        if source_uri is None:
+            return None
+
+        model_package_name = source_uri.split("/")[1]
+        return self.sagemaker_session.sagemaker_client.describe_model_package(
+            ModelPackageName=model_package_name
+        )
+
+    def endpoints(
```
Comment (staubhp) [inline src/sagemaker/lineage/action.py:345]: Missing return type

File: src/sagemaker/lineage/action.py [hunk 5]
```diff
@@ -250,3 +259,107 @@ def list(
...
+        self, direction: LineageQueryDirectionEnum = LineageQueryDirectionEnum.DESCENDANTS
+    ):
+        """Use a lineage query to retrieve all downstream contexts that use this action.
```
Comment (staubhp) [inline src/sagemaker/lineage/action.py:348]: "all downstream contexts" -> "endpoint contexts"

File: src/sagemaker/lineage/query.py [hunk 6]
```diff
@@ -203,11 +204,11 @@ def __init__(
     def _to_request_dict(self):
         """Convert the lineage filter to its API representation."""
         filter_request = {}
-        if self.entities:
```
Comment (staubhp) [inline src/sagemaker/lineage/query.py:206]: Woops. Good catch

Comment (shreyapandit) [reply src/sagemaker/lineage/query.py:206]: +1!

File: src/sagemaker/lineage/action.py (hunk 4 above)
Comment (yzhu0) [reply src/sagemaker/lineage/action.py:350]: > Missing return type

we have return types :)

File: src/sagemaker/lineage/action.py (hunk 3 above)
Comment (yzhu0) [reply src/sagemaker/lineage/action.py]: > This return type is wrong

updated

File: src/sagemaker/lineage/action.py (hunk 1 above)
Comment (yzhu0) [reply src/sagemaker/lineage/action.py]: > Either way, let's try to cut down on the code duplication if possible. Have a `_artifacts()` method that both of these call with the common code

To make it consistent with other methods, I updated to have `artifacts` with default both direction. The user can update the direction to upstream, downstream or both if they want to

File: src/sagemaker/lineage/action.py (hunk 4 above)
Comment (staubhp) [reply src/sagemaker/lineage/action.py:324]: Am I missing them? I mean `def endpoints() -> List[Context]:`

Comment (yzhu0) [reply src/sagemaker/lineage/action.py:325]: > Am I missing them? I mean `def endpoints() -> List[Context]:`

updated

**Review by staubhp — APPROVED** [review]

**Review by mufaddal-rohawala — APPROVED** [review]

**Review by ahsan-z-khan — APPROVED** [review]

**Pull request merged**